### PR TITLE
lease: grant consistent lease IDs

### PR DESCRIPTION
**Pull request by heyitsanthony**
When raft broadcasts a Grant to all nodes, all nodes must
agree on the same lease ID. Otherwise, attaching a key to
a lease will fail since the lease ID is node-dependent.

### CODE REVIEW

File: lease/lessor.go [hunk 1]
```diff
@@ -381,21 +382,19 @@ func (le *lessor) initAndRecover() {
 			expiry: forever,
 		}
 	}
-	tx.Unlock()
 
-	le.b.ForceCommit()
-}
+	// get last granted lease (if any) to seed grant values
+	_, vs = tx.UnsafeRange(leaseBucketName, lastLeaseKey, lastLeaseKey, 0)
+	if len(vs) != 0 {
```
Comment (gyuho) [inline lease/lessor.go]: `Just question. Would this always return 1 value `if len(vs) != 0` because we keep overwriting `lastLeaseKey`?

**Comment by gyuho** [comment]
LGTM. Defer to @xiang90 

Thanks,

File: lease/lessor.go (hunk 1 above)
Comment (heyitsanthony) [inline lease/lessor.go]: Yes, `len(vs)` will be at most 1.

Comment (gyuho) [inline lease/lessor.go]: got it. thanks

File: lease/lessor.go [hunk 2]
```diff
@@ -173,11 +168,17 @@ func (le *lessor) SetRangeDeleter(rd RangeDeleter) {
 // TODO: when lessor is under high load, it should give out lease
 // with longer TTL to reduce renew load.
 func (le *lessor) Grant(ttl int64) *Lease {
-	id := LeaseID(le.idgen.Next())
-
 	le.mu.Lock()
 	defer le.mu.Unlock()
 
+	// granted leases numbers are consistent via raft-pass through
+	// TODO: this bakes in the lease assignment values; upgrading
```
Comment (xiang90) [inline lease/lessor.go]: This approach seems OK. I feel the alternative is slightly better though. I think it is not hard to implement either. We can add ID here https://github.com/coreos/etcd/blob/master/etcdserver/etcdserverpb/rpc.proto#L263 and generate a unique ID here https://github.com/coreos/etcd/blob/master/etcdserver/api/v3rpc/lease.go#L35 before we send the request through raft.

I think if we are going to do the alternative, we'd better do it now.

Comment (heyitsanthony) [inline lease/lessor.go]: OK, I agree sending the lease id through raft is better; keeping the next lease id in sync is too delicate. Will change.

**Comment by heyitsanthony** [comment]
updated to send lease id over raft /cc @gyuho @xiang90

**Comment by gyuho** [comment]
lgtm

File: etcdserver/server.go [hunk 3]
```diff
@@ -167,9 +167,10 @@ type EtcdServer struct {
 
 	store store.Store
 
-	kv     dstorage.ConsistentWatchableKV
-	lessor lease.Lessor
-	be     backend.Backend
+	kv         dstorage.ConsistentWatchableKV
+	lessor     lease.Lessor
+	leaseIDGen *idutil.Generator
```
Comment (xiang90) [inline etcdserver/server.go]: We can actually reuse the reqIDGen or move the leaseIDGen to the API layer.

File: integration/v3_grpc_test.go [hunk 4]
```diff
@@ -1006,3 +1007,97 @@ func TestV3RangeRequest(t *testing.T) {
...
+	put := &pb.PutRequest{Key: []byte(key), Lease: lresp.ID}
+	if _, err := pb.NewKVClient(clus.RandConn()).Put(context.TODO(), put); err != nil {
+		return 0, err
+	}
+	return lresp.ID, nil
+}
+
+// testLeaseAction performs some action while holding a lease with an
+// attached key, then confirms the key is gone.
+func testLeaseAction(t *testing.T, act func(pb.LeaseClient, int64) error) {
```
Comment (xiang90) [inline integration/v3_grpc_test.go]: I feel this func name is not descriptive. The functionality of this is to create a leased key and then confirm leased key is removed. `testLeaseAction` sounds like we test the lease action itself.

**Comment by xiang90** [comment]
LGTM after fixing the issues mentioned above.

File: etcdserver/server.go (hunk 3 above)
Comment (heyitsanthony) [inline etcdserver/server.go]: ok reusing `reqIDGen`

File: integration/v3_grpc_test.go (hunk 4 above)
Comment (heyitsanthony) [inline integration/v3_grpc_test.go]: renamed to `testLeaseRemoveLeasedKey`

**Comment by heyitsanthony** [comment]
ok all fixed up; will merge on greenlight

**Pull request merged**